src/core/net/dns_client.cpp

```diff
@@ -257,9 +307,60 @@ Error Client::Response::FindServiceInfo(Section aSection, const Name &aName, Ser
 
 Error Client::AddressResponse::GetAddress(uint16_t aIndex, Ip6::Address &aAddress, uint32_t &aTtl) const
 {
-    return FindHostAddress(kAnswerSection, Name(*mQuery, kNameOffsetInQuery), aIndex, aAddress, aTtl);
+    Error error = kErrorNone;
+    Name  name(*mQuery, kNameOffsetInQuery);
+
+#if OPENTHREAD_CONFIG_DNS_CLIENT_NAT64_ENABLE
+
+    // If the response is for an IPv4 address query or if it is an
+    // IPv6 address query response with no IPv6 address but with an
+    // embedded IPv4 in its additional section, we read the IPv4
+    // address and translate it to an IPv6 address.
+
+    QueryInfo info;
+
+    info.ReadFrom(*mQuery);
+
+    if ((info.mQueryType == kIp4AddressQuery) || mHasOnlyEmbeddedIp4)
+    {
+        Section     section;
+        ARecord     aRecord;
+        Ip6::Prefix nat64Prefix;
+
+        SuccessOrExit(error = GetNat64Prefix(nat64Prefix));
+
+        section = (info.mQueryType == kIp4AddressQuery) ? kAnswerSection : kAdditionalDataSection;
+        SuccessOrExit(error = FindARecord(section, name, aIndex, aRecord));
+
+        aAddress.SetFromTranslatedIp4Address(nat64Prefix, aRecord.GetAddress());
```

Comment:
Looking at this now, the method name seems somewhat strange. `aRecord.GetAddress()` is a straight-up IPv4 address, but the method name indicates this is a "translated IPv4 address".

Reply:
How about  `SetByTranslatingIp4Address()`? (open to better suggestions as well).

Reply:
I will rename it to `SynthesizeFromIp4Address()`.

Reply:
SGTM.